src/webgpu/util/conversion.ts

```diff
@@ -755,8 +771,69 @@ export function TypeMat(cols: number, rows: number, elementType: ScalarType): Ma
   return ty;
 }
 
-/** Type is a ScalarType, VectorType, or MatrixType. */
-export type Type = ScalarType | VectorType | MatrixType;
+/** StructType describes the type of WGSL Struct. */
+export class StructType {
+  readonly name: string;
+  readonly elementTypes: Type[]; // Element types
+
+  constructor(name: string, types: Type[]) {
+    this.name = name;
+    this.elementTypes = types;
+  }
+
+  /**
+   * @returns a Struct constructed from the values read from the buffer at the
+   * given byte offset
+   */
+  public read(buf: Uint8Array, offset: number): Struct {
+    const elements: Value[] = this.elementTypes.map(e => {
+      const result = e.read(buf, offset);
+      offset += e.size;
+      // vec3s have a padding element
+      if (e instanceof VectorType) {
+        if (e.width === 3) {
+          offset += e.elementType.size;
+        }
+      }
+      return result;
+    });
+    return struct(this.name, ...elements);
+  }
+
+  public toString(): string {
+    return this.name;
+  }
+
+  public get size(): number {
+    return this.elementTypes.reduce((sum, cur) => {
+      sum += cur.size;
+      // vec3s have a padding element
+      if (cur instanceof VectorType) {
+        if (cur.width === 3) {
+          sum += cur.elementType.size;
```

Comment:
The same here, we should compute the size of structure by the last member's position + it's size and round up to the maximum alignment of structure's members. [Spec](https://gpuweb.github.io/gpuweb/wgsl/#example-fc0bb4df:~:text=roundUp(AlignOf(S)%2C%20justPastLastMember)%0A%0Awhere%20justPastLastMember%20%3D%20OffsetOfMember(S%2CN)%20%2B%20SizeOfMember(S%2CN))

So, the size of
``` rust
struct A {                            //            align(16) size(16)
    a: vec3<f32>,                     // offset(0)  align(16) size(12)
    b: f32,                           // offset(12) align(4)  size(4)
}
```
would be 16 rather than 20.